benchmarks/CompilerServiceMemoryAnalysisRunner/Program.fs

```diff
@@ -0,0 +1,115 @@
+﻿open System.Diagnostics
+open System.Collections.Generic
+open System.Threading.Tasks
+open Microsoft.Diagnostics.Runtime
+
+// WARNING: This type name can change.
+let CompilerServiceProjectCacheTypeName = "Internal.Utilities.Collections.MruCache<FSharp.Compiler.AbstractIL.Internal.Library+CompilationThreadToken,FSharp.Compiler.SourceCodeServices.FSharpProjectOptions,System.Tuple<Microsoft.FSharp.Core.FSharpOption<FSharp.Compiler.IncrementalBuilder>,FSharp.Compiler.SourceCodeServices.FSharpErrorInfo[],System.IDisposable>>"
+
+[<RequireQualifiedAccess>]
+type ClrInstance =
+    | Object of ClrObject
+    | ValueClass of ClrValueClass
+
+let rec getFullMemorySize (instance: ClrInstance) (objSet: HashSet<uint64>) =
+    match instance with
+    | ClrInstance.Object(clrInstance) ->
+
+        if clrInstance.IsNull || not (objSet.Add clrInstance.Address) then 0
+        else
+
+        let instanceFieldFullMemorySize =
+            clrInstance.Type.Fields
```

Comment:
Can be
```fsharp
            clrInstance.Type.Fields
            |> Seq.sumBy (fun clrInstanceField ->
                if clrInstanceField.Type.IsValueClass then
                    let clrValueClass = clrInstance.GetValueClassField(clrInstanceField.Name)
                    getFullMemorySize (ClrInstance.ValueClass clrValueClass) objSet
                elif clrInstanceField.Type.IsPrimitive then
                    clrInstanceField.Type.BaseSize
                else
                    let clrObject = clrInstance.GetObjectField(clrInstanceField.Name)
                    getFullMemorySize (ClrInstance.Object clrObject) objSet
            )
```

Reply:
Yup, I also need to hoist some of this out as I got duplicate code.